pkg/storage/kubernetes/ipam.go

```diff
@@ -2,31 +2,168 @@ package kubernetes
+				return nil, err
+			}
+		}
+	}
+	return lease, nil
+}
+
+func acquireLease(ipam *KubernetesIPAM, ipamConf *whereaboutstypes.IPAMConfig) error {
+	var lease *coordinationv1.Lease
+	var err error
+
+	backoff := int64(ipamConf.Backoff)
+
+	clientID := fmt.Sprintf("%s/%s", ipamConf.PodNamespace, ipamConf.PodName)
+
+	leaseName := strings.ReplaceAll(ipamConf.Range, ":", "-")
+	leaseName = strings.ReplaceAll(leaseName, "/", "-")
+	logging.Debugf("client %s, trying to acquire lease %s", clientID,
+		leasePrefix+leaseName)
+
+	ctx := context.Background()
+
+	for {
+		lease, err = getLease(ctx, ipam, leaseName)
+		if err != nil {
+			return err
+		}
+		isLeaseExpired := false
+		if lease.Spec.HolderIdentity != nil {
+			acquireTime := lease.Spec.AcquireTime
+			leaseDuration := *lease.Spec.LeaseDurationSeconds
+
+			if int32(time.Since(acquireTime.Time).Seconds()) >= leaseDuration {
+				logging.Debugf("lease %s held by %s expired, acquireTime: %v",
+					lease.Name, *lease.Spec.HolderIdentity, acquireTime)
+				isLeaseExpired = true
+			}
+		}
+		// try to acquire the lease
+		if lease.Spec.HolderIdentity == nil || isLeaseExpired {
+			lease.Spec.AcquireTime = &metav1.MicroTime{Time: time.Now()}
+			*lease.Spec.LeaseTransitions = *lease.Spec.LeaseTransitions + 1
+			lease.Spec.HolderIdentity = &clientID
+
+			err = ipam.client.Update(ctx, lease)
+			if err == nil {
+				break
+			}
+			if !errors.IsConflict(err) {
+				logging.Errorf("error acquiring lease %s by client %s, error: %s", lease.Name, clientID, err)
+				return err
+			}
+		}
+		interval, _ := rand.Int(rand.Reader, big.NewInt(backoff))
+		sleepTime := time.Duration(int64(10) + interval.Int64())
+		logging.Debugf("client %s, sleeping for %d ms for lease %s", clientID, sleepTime, lease.Name)
+		time.Sleep(sleepTime * time.Millisecond)
+	}
+	if *lease.Spec.HolderIdentity != clientID {
+		return fmt.Errorf("failed to acquire lease %s after waiting, client: %s",
+			lease.Name, clientID)
+	}
+	logging.Debugf("lease %s acquired, client: %s", lease.Name, clientID)
+	return nil
+}
+
+func releaseLease(ipam *KubernetesIPAM, ipamConf *whereaboutstypes.IPAMConfig) error {
+	leaseName := strings.ReplaceAll(ipamConf.Range, ":", "-")
+	leaseName = strings.ReplaceAll(leaseName, "/", "-")
+
+	clientID := fmt.Sprintf("%s/%s", ipamConf.PodNamespace, ipamConf.PodName)
+
+	key := client.ObjectKey{
+		Name:      fmt.Sprintf("%s-%s", leasePrefix, leaseName),
+		Namespace: ipam.namespace,
+	}
+	lease := &coordinationv1.Lease{}
+	ctx := context.Background()
+	err := ipam.client.Get(ctx, key, lease)
+	if err != nil {
+		logging.Errorf("client %s, failed to get lease %s, err: %s", clientID, key.Name, err)
+		return err
+	}
+	// check if this client is the owner of the lease and update
+	if lease.Spec.HolderIdentity == nil || *lease.Spec.HolderIdentity != clientID {
+		return fmt.Errorf("client %s, lease %s already cleared or held by someone else",
+			clientID, lease.Name)
+	}
```

Comment:
This would merit yet another helper - `func isLeaseOwnedBy(clientID string) bool {...}` .

Reply:
Done